### DISCUSSION TITLE
Type of request.client

### DISCUSSION THREAD

**Discussion by andrew222651** [encode/starlette#2244] [Q&A] [answered]
The docs [say](https://www.starlette.io/requests/#:~:text=Either%20item%20in%20the%20tuple%20may%20be%20None.) that  `Request.client` is a 2-tuple and either component may be `None`, but [in the code](https://github.com/encode/starlette/blob/a8b8856ce393a82ab4a714131085dbc4d658e34d/starlette/requests.py#L138) the type of `Request.client` is `Optional[Address]` where `Address` is
```python
class Address(typing.NamedTuple):
    host: str
    port: int
```
So if the code reflects the actual behavior, should the docs be reworded to say `Request.client` itself may be `None` but its components aren't `None`?

**Comment by Kludex** [comment, marked as answer]
I think the documentation is wrong. I remember that type hint being fixed to match what is on the ASGI spec: https://asgi.readthedocs.io/en/latest/specs/www.html#http-connection-scope.

PR welcome to fix the documentation - please also confirm what I said (a blame on the `Address` would likely point you to a PR).

**Reply by andrew222651** [reply]
Do you know in what circumstances `Request.client` is `None`?

**Reply by iamdbychkov** [reply]
The case of `request.client` being `None` (that I know of) is when ASGI server didn't put this information in a scope.

For e.g. uvicorn listening on a UNIX socket won't give you client information until you put `"forwarded_allow_ips": "*"` in a settings, thus `request.client` would return `None`
